api/src/main/java/org/apache/iceberg/IncrementalScan.java

```diff
@@ -21,6 +21,23 @@
 /** API for configuring an incremental scan. */
 public interface IncrementalScan<ThisT, T extends ScanTask, G extends ScanTaskGroup<T>>
     extends Scan<ThisT, T, G> {
+
+  /**
+   * Instructs this scan to look for changes starting from a particular snapshot (inclusive).
+   *
+   * <p>If the start snapshot is not configured, it is defaulted to the oldest ancestor of the end
+   * snapshot (inclusive).
+   *
+   * @param fromSnapshotId the start snapshot ID (inclusive)
+   * @param referenceName the ref used
+   * @return this for method chaining
+   * @throws IllegalArgumentException if the start snapshot is not an ancestor of the end snapshot
+   */
+  default ThisT fromSnapshotInclusive(long fromSnapshotId, String referenceName) {
```

Comment:
I prefer to use `to` to point to a specific snapshot. 
For example, we want to point to a historical snapshot of the branch rather than the latest snapshot of the branch:
```
 table.newIncrementalScan()
      .fromSnapshotExclusive(lastProcessedId)
      .toSnapshot(1)
      .toBranch("some-branch-name")
      .planTasks()
```
Personally, I think it looks strange.

Considering your suggestion, I prefer to add some methods: `toTag(String tagName)` `fromTagInclusive(String tagName)` and `fromTagExclusive(String tagName)`, which points to a specific snapshot reference. `useBranch(String branchName)`, the branch we're going to use. 
We will have the following example usage:
```
 table.newIncrementalScan()
      .fromTagInclusive(tag1)
      .toTag(tag2)
      .useBranch("some-branch-name")
      .planTasks()
```

Reply:
Follow along @nastra 's suggestion, we can have overload methods via argument type (id vs reference). Id (long) or reference (string) are just different ways to specify the snapshot. it is good that current methods are called `fromSnapshot`.

```
ThisT fromSnapshotInclusive(long fromSnapshotId);
ThisT fromSnapshotInclusive(String fromSnapshotRef);
```

Reply:
`fromSnapshotInclusive(String fromSnapshotRef)` is fine for me, but one thing I want to clarify is that if we pass in a tag, then it represents the snapshot of the tag binding, but what does it mean if it's a branch? Use the current latest snapshot of the branch and switch the branch to `fromSnapshotRef`? Or it not allowed to use branch?

Reply:
We assume that the `fromSnapshotInclusive(String fromSnapshotRef)` is just use the snapshot that the tag binding, or the latest snapshot of the branch.

Currently have the following methods:
```
ThisT fromSnapshotInclusive(String fromSnapshotRef) // Use the snapshot that tag bound with, or the latest of the branch.

ThisT fromSnapshotExclusive(String fromSnapshotRef); // Similar to the above

ThisT toSnapshot(String fromSnapshotRef); // The end snapshot, tag bound to, or the latest snapshot of the branch.

ThisT useBranch(String branchName); // switches the branch to the specified one.
```
We can have the following scenarios

```
table.newIncrementalScan()
      .fromSnapshotInclusive(snapshoitId or ref)
      .useBranch(branch1)
      .planTasks()
```
Specify the start snapshsot or ref and switch the branch to branch1.  
Can be used in streaming tasks

```
table.newIncrementalScan()
      .fromSnapshotInclusive(snapshoitId or ref)
      .toSnapshot(snapshoitId or ref)
      .planTasks()
```
Specifies the start and end snapshot or ref. 
Can be used in batch tasks.

Reply:
`fromSnapshot` can't be a branch name. it probably has to be a tag or id. `toSnapshot` can be a branch which indicates the latest snapshot of the branch.

Reply:
@nastra, the distinction between a branch and a tag here is that a tag does not change. It doesn't make sense to read incremental changes from a tag because it is a fixed point in time.

Reply:
> @nastra, the distinction between a branch and a tag here is that a tag does not change. It doesn't make sense to read incremental changes from a tag because it is a fixed point in time. So I think it's okay to use `toBranch` instead of `toRef`, although I'm not sure it's actually worth the distinction since the other API uses `ref`.

my main point was purely on the naming of the API (`toBranch` vs `toRef`). It seems to be we could have `toRef()`, which would perform a check to make sure the given ref is an actual branch

Reply:
@rdblue, I know what you mean, the tag does point to a fixed snapshot, but as I said above, can we use it to replace a snapshot ID when we are running in batch mode? In this case, we are not reading the increment changes from a tag, but from a particular start snapshot ID  to the fixed snapshot ID which the TAG points. Using TAG is more friendly than using snapshot ids.
For example:
```
table.newIncrementalScan()
      .fromSnapshotInclusive(snapshoitId or tag)
      .toSnapshot(snapshoitId or tag)
      .planTasks()
```

Reply:
>  It doesn't make sense to read incremental changes from a tag because it is a fixed point in time.

This is an interesting point. A tag is just a name pointing to some snapshot in the past. There are two possibilities.

* the tag is outside the snapshot retention window (e.g. 7 days). Then this is an invalid case for incremental scan
* the tag is still within the snapshot retention window (e.g. 7 days). Then I think it is fine to have increment scan with `fromSnapshot(String tag)`. E.g., maybe the Iceberg table has daily tags for bookkeeping of good rewind points for easier backfill.

We can allow `fromSnapshot(String tag)` but add a validation?

@rdblue what do you think?

Reply:
Agree with @stevenzwu.
Yes, tag is a fixed point in time, but when using it for incremental read, we can think of it semantically the same as using `fromSnapshot(Long snapshotId)`.
Just like the @stevenzwu's example, I have  daily tags(`20220101` `20220102`), If I want to read the incremental data from `20220102` to the current., so I can use `fromSnapshotExclusive("20220102")`:
```
table.newIncrementalScan()
      .fromSnapshotExclusive("20220102")
      .planTasks()
```
Another way is to use snapshot time to find the snapshot id first, but sometimes that doesn't work. For example, we can generate tags based on the event time of the data, or we tag the snapshot only after the application has completed. The application may finish at 3:00 2022/01/02, and tag the newly generated snapshot as `20220102`. If we use snapshot time `2022-01-02 00:00:00` to find the snapshot ID, incorrect incremental data will be return.

cc @rdblue